credentials/sts/sts.go

```diff
@@ -0,0 +1,427 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package sts implements call credentials using STS (Security Token Service) as
+// defined in https://tools.ietf.org/html/rfc8693.
+//
+// Experimental
+//
+// Notice: All APIs in this package are experimental and may be changed or
+// removed in a later release.
+package sts
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	grpcbackoff "google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/backoff"
+)
+
+const (
+	// HTTP request timeout set on the http.Client used to make STS requests.
+	stsRequestTimeout = 5 * time.Second
+	// If lifetime left in a cached token is lesser than this value, we fetch a
+	// new one instead of returning the current one.
+	minCachedTokenLifetime = 300 * time.Second
+
+	tokenExchangeGrantType    = "urn:ietf:params:oauth:grant-type:token-exchange"
+	defaultCloudPlatformScope = "https://www.googleapis.com/auth/cloud-platform"
+)
+
+// For overriding in tests.
+var (
+	loadSystemCertPool   = x509.SystemCertPool
+	makeBackoffStrategy  = defaultBackoffStrategy
+	makeHTTPDoer         = makeHTTPClient
+	readSubjectTokenFrom = ioutil.ReadFile
+	readActorTokenFrom   = ioutil.ReadFile
+)
+
+// Options configures the parameters used for an STS based token exchange.
+type Options struct {
+	// TokenExchangeServiceURI is the address of the server which implements STS
+	// token exchange functionality.
+	TokenExchangeServiceURI string // Required.
+
+	// Resource is a URI that indicates the target service or resource where the
+	// client intends to use the requested security token.
+	Resource string // Optional.
```

Comment:
Should we document what the default behavior is if these optional fields are unset?

Reply:
Ahh.  It wasn't apparent to me that these were options sent in a request when I was first reading the code, vs. options used by a local library.  So this is a sub-set of `RequestParameters` then?  Why have both?

Reply:
They are not exactly similar. Apart from the fields `TokenExchangeServiceURI` and `GrantType` being present in only one of the structs, the more important difference is that in `Options`, we have `SubjectTokenPath` and `ActorTokenPath` which point to files which contain the appropriate tokens, while in `RequestParameters` we have `SubjectToken` and `ActorToken` which contain the actual tokens.

Reply:
Should we unexport `RequestParameters` if they aren't needed by users of the package?

Reply:
Done.